File: app/src/main/java/com/example/admin/mt_lab1/MainActivity.java
package com.example.admin.mt_lab1;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.Toast;


public class MainActivity extends AppCompatActivity {

    Button[][] fields = new Button[3][3];
    TableLayout layout;
    int leadCounter = 0;
    boolean currentPlayer1 = true;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        layout = findViewById(R.id.table);

        buildGame();
    }

    private void buildGame() {

        for(int i = 0; i < 3; i++){
            TableRow row = new TableRow(this);
            for(int j = 0; j < 3; j++){
                final Button button = new Button(this);
                fields[i][j] = button;
                row.addView(button, new TableRow.LayoutParams(360, 560));
                fields[i][j].setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {

                        if(currentPlayer1){
                            button.setText("X");
                            button.setClickable(false);
                            leadCounter++;
                            checkVictory();
                            currentPlayer1 = !currentPlayer1;
                        }else{
                            button.setText("O");
                            button.setClickable(false);
                            leadCounter++;
                            checkVictory();
                            currentPlayer1 = !currentPlayer1;
                        }
                    }
                });
            }

            layout.addView(row, new TableLayout.LayoutParams(TableLayout.LayoutParams.MATCH_PARENT, TableLayout.LayoutParams.MATCH_PARENT));
        }
    }

    private void checkVictory(){

        if(fields[0][0].getText() == "X" && fields[0][1].getText() == "X" && fields[0][2].getText() == "X"
        || fields[1][0].getText() == "X" && fields[1][1].getText() == "X" && fields[1][2].getText() == "X"
        || fields[2][0].getText() == "X" && fields[2][1].getText() == "X" && fields[2][2].getText() == "X"
        || fields[0][0].getText() == "X" && fields[1][0].getText() == "X" && fields[2][0].getText() == "X"
        || fields[0][1].getText() == "X" && fields[1][1].getText() == "X" && fields[2][1].getText() == "X"
        || fields[0][2].getText() == "X" && fields[1][2].getText() == "X" && fields[2][2].getText() == "X"
        || fields[0][0].getText() == "X" && fields[1][1].getText() == "X" && fields[2][2].getText() == "X"
        || fields[0][2].getText() == "X" && fields[1][1].getText() == "X" && fields[2][0].getText() == "X"){

            Toast.makeText(this, "Player 1 (X) won!", Toast.LENGTH_SHORT).show();
            refresh();
        }
        else if(fields[0][0].getText() == "O" && fields[0][1].getText() == "O" && fields[0][2].getText() == "O"
             || fields[1][0].getText() == "O" && fields[1][1].getText() == "O" && fields[1][2].getText() == "O"
             || fields[2][0].getText() == "O" && fields[2][1].getText() == "O" && fields[2][2].getText() == "O"
             || fields[0][0].getText() == "O" && fields[1][0].getText() == "O" && fields[2][0].getText() == "O"
             || fields[0][1].getText() == "O" && fields[1][1].getText() == "O" && fields[2][1].getText() == "O"
             || fields[0][2].getText() == "O" && fields[1][2].getText() == "O" && fields[2][2].getText() == "O"
             || fields[0][0].getText() == "O" && fields[1][1].getText() == "O" && fields[2][2].getText() == "O"
             || fields[0][2].getText() == "O" && fields[1][1].getText() == "O" && fields[2][0].getText() == "O"){

            Toast.makeText(this, "Player 2 (O) won!", Toast.LENGTH_SHORT).show();
            refresh();
        }
        else if(leadCounter == 9){
            Toast.makeText(this, "Draw!", Toast.LENGTH_SHORT).show();
            refresh();
        }
    }

    private void refresh(){

        for(int i = 0; i < 3; i++){
            for(int j = 0; j < 3; j++){

                fields[i][j].setText("");
                fields[i][j].setClickable(true);
                leadCounter = 0;
            }
        }
    }
}

File: app/src/main/java/com/example/android/retrofitgsonplusroom/MainActivity.kt
package com.example.android.retrofitgsonplusroom

import android.content.Context
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import com.example.android.retrofitgsonplusroom.api.Api
import com.example.android.retrofitgsonplusroom.model.Locations
import com.example.android.retrofitgsonplusroom.room.LocationsDatabase
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_main.*
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory


class MainActivity : AppCompatActivity() {

    lateinit var locationslist: Locations
    lateinit var context: Context
    lateinit var mydb: LocationsDatabase
    var counter = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        context = applicationContext
        mydb = LocationsDatabase.getInstance(context)

        //Next Button
        button_next.setOnClickListener {
            counter++
            if (counter >= locationslist.residents.size - 1) {
                counter = 0
            }
            textview.setText(locationslist.residents[counter])
            Toast.makeText(applicationContext, "Next", Toast.LENGTH_LONG).show()
            Single.fromCallable({mydb.locationsDao().insertCurrent(locationslist)}).subscribeOn(Schedulers.io())
                .subscribe({
                    Log.i("insertCurrent", "+++");
                }, {
                    Log.i("Error ", "---");
                })
        }

        //Previous Button
        button_previous.setOnClickListener {
            Toast.makeText(applicationContext, "Previous", Toast.LENGTH_LONG).show()
            counter--
            if (counter <= 0) {
                counter = locationslist.residents.size - 1
            }
            textview.setText(locationslist.residents[counter])
            Single.fromCallable({mydb.locationsDao().insertCurrent(locationslist)}).subscribeOn(Schedulers.io())
                .subscribe({
                    Log.i("insertCurrent ", "+++");
                }, {
                    Log.i("Error ", "---");
                })
        }

        //Room Button
        button_room.setOnClickListener {
            Toast.makeText(applicationContext, "Room", Toast.LENGTH_LONG).show()

            mydb.locationsDao().getAll()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    Log.d("getAll", "dogs image urls" + it.toString())
                },{
                    Log.d("Error:getAll", "lols")
                })
        }

        textview.setText("Zaglushka")

        //Retrofit
        val retrofit = Retrofit.Builder()
            .baseUrl("https://rickandmortyapi.com/api/location/")
            .addConverterFactory(GsonConverterFactory.create())
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .build()
        val service = retrofit.create(Api::class.java)
        service.getId()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    locationslist = it
                    textview.setText(locationslist.residents[1].toString())
                    Log.i("StringMessage", it.residents.toString())
                }, {
                    Log.i("Subscribe ", "Error")
                })
    }
}
